File: DynamicKey/AgoraDynamicKey/nodejs/src/DynamicKey.js
/**
* Server side. Generate Client key.
*/

var crypto = require('crypto');

var encodeHMac = function(key, message) {
    return crypto.createHmac('sha1', key).update(message).digest('hex');
};

module.exports.generate = function(appID, appCertificate, channelName, unixTs, randomInt) {
    channelName=channelName.toString();
    var unixTsStr = ("0000000000" + unixTs).substring(String(unixTs).length);
    var rndTxt = randomInt.toString(16);
    var randomIntStr = ("00000000" + rndTxt).substring(rndTxt.length);
    var sign = generateSignature(appID, appCertificate, channelName, unixTsStr, randomIntStr);
    return sign + appID + unixTsStr + randomIntStr;
};

var generateSignature = function(appID, appCertificate, channelName, unixTsStr, randomIntStr) {
    var buffer = Buffer.concat([new Buffer(appID), new Buffer(unixTsStr), new Buffer(randomIntStr), new Buffer(channelName)]);
    var sign = encodeHMac(appCertificate, buffer);
    return sign.toString('hex');
};

File: DynamicKey/AgoraDynamicKey/go/src/DynamicKey/DynamicKey.go
package DynamicKey

import (
	"crypto/hmac"
	"crypto/sha1"
	"encoding/hex"
	"fmt"
	"strings"
)

func Generate(appID, appCertificate, channelName string, unixTs, randomInt uint32) (string) {
    return generateDynamicKey(appID, appCertificate, channelName, unixTs, randomInt)
}

func generateDynamicKey(appID, appCertificate, channelName string, unixTs, randomInt uint32) (string) {
	unixTsStr := fmt.Sprintf("%010d", unixTs)
	randomIntStr := fmt.Sprintf("%08x", randomInt)
	signature := generateSignature(appID, appCertificate, channelName, unixTsStr, randomIntStr)
	buffer := strings.Join([]string{signature, appID, unixTsStr, randomIntStr}, "")
	return buffer
}

func generateSignature(appID, appCertificate, channelName, unixTsStr, randomIntStr string) (string) {
	buffer := strings.Join([]string{appID, unixTsStr, randomIntStr, channelName}, "")
	signature := hmac.New(sha1.New, []byte(appCertificate))
	signature.Write([]byte(buffer))
	return hex.EncodeToString(signature.Sum(nil))
}
